flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/UnboundedEventTimeOverProcessFunction.scala

```diff
@@ -0,0 +1,283 @@
+import org.apache.flink.api.common.typeutils.base.StringSerializer
+import org.apache.flink.api.java.functions.KeySelector
+import org.apache.flink.api.java.tuple.Tuple
+import org.apache.flink.core.memory.{DataInputViewStreamWrapper, DataOutputViewStreamWrapper}
+import org.apache.flink.runtime.state.{FunctionInitializationContext, FunctionSnapshotContext}
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction
+import org.apache.flink.streaming.api.operators.TimestampedCollector
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow
+import org.apache.flink.table.functions.{Accumulator, AggregateFunction}
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+  * A ProcessFunction to support unbounded event-time over-window
+  *
+  * @param aggregates the aggregate functions
+  * @param aggFields  the filed index which the aggregate functions use
+  * @param forwardedFieldCount the input fields count
+  * @param interMediateType the intermediate row tye which the state saved
+  * @param keySelector the keyselector
+  * @param keyType     the key type
+  *
+  */
+class UnboundedEventTimeOverProcessFunction(
+    private val aggregates: Array[AggregateFunction[_]],
+    private val aggFields: Array[Int],
+    private val forwardedFieldCount: Int,
+    private val interMediateType: TypeInformation[Row],
+    private val keySelector: KeySelector[Row, Tuple],
+    private val keyType: TypeInformation[Tuple])
+  extends ProcessFunction[Row, Row]
+  with CheckpointedFunction{
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var output: Row = _
+  private var state: MapState[TimeWindow, Row] = _
+  private val aggregateWithIndex: Array[(AggregateFunction[_], Int)] = aggregates.zipWithIndex
+
+  /** Sorted list per key for choose the recent result and the records need retraction **/
+  private val timeSectionsMap: java.util.HashMap[Tuple, java.util.LinkedList[TimeWindow]] =
+        new java.util.HashMap[Tuple, java.util.LinkedList[TimeWindow]]
+
+  /** For store timeSectionsMap **/
+  private var timeSectionsState: ListState[String] = _
+  private var inputKeySerializer: TypeSerializer[Tuple] = _
+  private var timeSerializer: TypeSerializer[TimeWindow] = _
+
+  override def open(config: Configuration) {
+    output = new Row(forwardedFieldCount + aggregates.length)
+    val valueSerializer: TypeSerializer[Row] =
+      interMediateType.createSerializer(getRuntimeContext.getExecutionConfig)
+    timeSerializer = new TimeWindow.Serializer
+    val stateDescriptor: MapStateDescriptor[TimeWindow, Row] =
+      new MapStateDescriptor[TimeWindow, Row]("rowtimeoverstate", timeSerializer, valueSerializer)
+    inputKeySerializer = keyType.createSerializer(getRuntimeContext.getExecutionConfig)
+    state = getRuntimeContext.getMapState[TimeWindow, Row](stateDescriptor)
+  }
+
+  override def processElement(
```

Comment:
Waiting for watermarks is the only way to handle out-of-order data without sending out retractions. Since, retractions are not supported yet, it is the only viable option. Otherwise, we would need to drop all records which have not monotonically increasing timestamps.